### PR TITLE
IGNITE-19231 Change thread pool for metastore raft group

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-19231

### CODE REVIEW

File: modules/raft-api/src/main/java/org/apache/ignite/internal/raft/RaftNodeDisruptorConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+     * @param threadPostfix Postfix the name of the disruptor threads.
+     * @param stripes Number of disruptor stripes.
+     */
+    public RaftNodeDisruptorConfiguration(String threadPostfix, int stripes) {
+        this.threadPostfix = threadPostfix;
+        this.stripes = stripes;
+    }
+
+    /**
+     * Return postfix the name of the disruptor threads.
```
Comment (rpuch) [inline modules/raft-api/src/main/java/org/apache/ignite/internal/raft/RaftNodeDisruptorConfiguration.java:40]:
```suggestion
     * Return Disruptor threads' name postfix.
```

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/PartitionListener.java [hunk 2]
```diff
@@ -175,6 +176,14 @@ public void onWrite(Iterator<CommandClosure<WriteCommand>> iterator) {
                 clo.result(null);
             } catch (IgniteInternalException e) {
                 clo.result(e);
+            } catch (CompletionException e) {
+                clo.result(e.getCause());
+            } catch (Throwable t) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/PartitionListener.java:181]: Of course I forgot to do that, thanks!

File: modules/raft/src/integrationTest/java/org/apache/ignite/raft/jraft/core/ItNodeTest.java [hunk 3]
```diff
@@ -3976,7 +3976,7 @@ private RaftGroupService createService(String groupId, TestPeer peer, NodeOption
 
         clusterService.start();
 
-        var service = new RaftGroupService(groupId, peer.getPeerId(), nodeOptions, rpcServer, nodeManager) {
+        var service = new RaftGroupService(groupId, peer.getPeerId(), nodeOptions, rpcServer, nodeManager, null) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/raft/src/integrationTest/java/org/apache/ignite/raft/jraft/core/ItNodeTest.java:3979]: Why not.

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftServer.java [hunk 4]
```diff
@@ -76,6 +77,26 @@ boolean startRaftNode(
...
 
+    /**
+     * Starts a Raft group bound to this cluster node.
+     *
+     * @param nodeId Raft node ID.
+     * @param configuration Raft configuration.
+     * @param evLsnr Listener for group membership and other events.
+     * @param lsnr Listener for state machine events.
+     * @param groupOptions Options to apply to the group.
+     * @param ownFsmCallerExecutorDisruptorConfig Configuration own (not shared) striped disruptor for FSMCaller service of raft node.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftServer.java:88]: fix it

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java [hunk 5]
```diff
@@ -225,6 +227,9 @@ public class NodeImpl implements Node, RaftServerService {
      */
     private volatile int electionTimeoutCounter;
 
+    /** Configuration own striped disruptor for FSMCaller service of raft node, {@code null} means use shared disruptor. */
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java:230]: fix it

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/JRaftUtils.java [hunk 6]
```diff
@@ -54,7 +54,7 @@ public final class JRaftUtils {
      * @return true if bootstrap success
      */
     public static boolean bootstrap(final BootstrapOptions opts) throws InterruptedException {
-        final NodeImpl node = new NodeImpl("bootstrap", new PeerId("127.0.0.1", 0));
+        final NodeImpl node = new NodeImpl("bootstrap", new PeerId("127.0.0.1", 0), null);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/JRaftUtils.java:57]: Why not.

File: modules/raft-api/src/main/java/org/apache/ignite/internal/raft/RaftNodeDisruptorConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+ */
+public class RaftNodeDisruptorConfiguration {
+    private final String threadPostfix;
+
+    private final int stripes;
+
+    /**
+     * Constructor.
+     *
+     * @param threadPostfix Postfix the name of the disruptor threads.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/raft-api/src/main/java/org/apache/ignite/internal/raft/RaftNodeDisruptorConfiguration.java:31]: fix it

File: modules/raft-api/src/main/java/org/apache/ignite/internal/raft/RaftNodeDisruptorConfiguration.java (hunk 1 above)
Comment (tkalkirill) [reply modules/raft-api/src/main/java/org/apache/ignite/internal/raft/RaftNodeDisruptorConfiguration.java:40]: Fix it

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java [hunk 8]
```diff
@@ -235,7 +236,36 @@ public <T extends RaftGroupService> CompletableFuture<T> startRaftGroupNode(
...
+        }
+    }
+
+    @Override
+    public CompletableFuture<RaftGroupService> startRaftGroupNode(
+            RaftNodeId nodeId,
+            PeersAndLearners configuration,
+            RaftGroupListener lsnr,
+            RaftGroupEventsListener eventsLsnr,
+            RaftNodeDisruptorConfiguration ownFsmCallerExecutorDisruptorConfig
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java:251]: Look, `RaftGroupOptions` is in module `ignite-raft` and `RaftManager` (as well as `RaftNodeDisruptorConfiguration`) are in module `ignite-raft-api`, I don't want to inflate the amount of work in this ticket, let's not do it in **IGNITE-18273**. 

IGNITE-18273 already indicates what needs to be done and within it, we will move the configuration, I will mention this in it.

[the remark this answers is not in the document]
Comment (rpuch) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java:251]: At least `RaftServer` is in the same module as `RaftGroupOptions`, so it seems that it is possible to pass `RaftNodeDisruptorConfiguration` to `RaftServer` inside of `RaftGroupOptions` easily. Does it take a lot of work? On the other hand, `if this is done, `RaftServer` will become simpler right now.

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/RaftGroupService.java [hunk 9]
```diff
@@ -76,26 +76,51 @@ public class RaftGroupService {
...
+         */
+        public RaftGroupService(
+                final String groupId,
+                final PeerId serverId,
+                final NodeOptions nodeOptions,
+                final RpcServer rpcServer,
+                final NodeManager nodeManager,
+                final RaftNodeDisruptorConfiguration ownFsmCallerExecutorDisruptorConfig
+        ) {
+            super();
```
Comment (rpuch) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/RaftGroupService.java:112]: Sorry, I did not express this clearly: the idea was to reuse one constructor inside another. Looks like the only thing that prevents this is the `assert`, but it doesn't seem to be too important. By removing it, we can remove the duplication. WDYT?

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java [hunk 10]
```diff
@@ -571,9 +570,32 @@ public NodeImpl(
         this.confCtx = new ConfigurationCtx(this);
         this.wakingCandidate = null;
         this.applyCommittedFuture = new CompletableFuture<>();
-        this.ownFsmCallerExecutorDisruptorConfig = ownFsmCallerExecutorDisruptorConfig;
+        this.ownFsmCallerExecutorDisruptorConfig = null;
     }
 
+        public NodeImpl(
```
Comment (rpuch) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java:576]: Same thing: let's reuse one constructor in the other. This seems to be especially important in this class, as the constructor has some non-trivial logic, not just assignments to fields, and it seems scary to duplicate this logic.

Comment (tkalkirill) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java:576]: Fix it

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/RaftGroupService.java (hunk 9 above)
Comment (tkalkirill) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/RaftGroupService.java:112]: Fix it.

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java:251]: I'll try to fix it.

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java [hunk 11]
```diff
@@ -37,6 +39,9 @@ public class RaftGroupOptions {
     /** Raft meta storage factory. */
     private RaftMetaStorageFactory raftMetaStorageFactory;
 
+    /** Configuration own striped disruptor for FSMCaller service of raft node, {@code null} means use shared disruptor. */
```
Comment (rpuch) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java:42]:
```suggestion
    /** Configuration of own striped disruptor for FSMCaller service of raft node, {@code null} means use shared disruptor. */
```

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java [hunk 12]
```diff
@@ -126,4 +131,20 @@ public RaftGroupOptions raftMetaStorageFactory(RaftMetaStorageFactory raftMetaSt
 
         return this;
     }
+
+    /**
+     * Returns configuration own striped disruptor for FSMCaller service of raft node, {@code null} means use shared disruptor.
```
Comment (rpuch) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java:136]:
```suggestion
     * Returns configuration of own striped disruptor for FSMCaller service of raft node, {@code null} means use shared disruptor.
```

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java [hunk 13]
```diff
@@ -126,4 +131,20 @@ public RaftGroupOptions raftMetaStorageFactory(RaftMetaStorageFactory raftMetaSt
...
+     */
+    public @Nullable RaftNodeDisruptorConfiguration ownFsmCallerExecutorDisruptorConfig() {
+        return ownFsmCallerExecutorDisruptorConfig;
+    }
+
+    /**
+     * Sets configuration own striped disruptor for FSMCaller service of raft node.
```
Comment (rpuch) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java:143]:
```suggestion
     * Sets configuration of own striped disruptor for FSMCaller service of raft node.
```

**Review by rpuch — APPROVED** [review]

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java (hunk 11 above)
Comment (tkalkirill) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java:42]: Fix it

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java (hunk 13 above)
Comment (tkalkirill) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java:143]: fix it

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java (hunk 12 above)
Comment (tkalkirill) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupOptions.java:136]: fix it

**Pull request merged**